app/lib/logstore_standard.js

```diff
@@ -267,96 +292,104 @@ module.exports = function(module, action){
+                        // Given defining_attributes('foo') => ["email"],
+                        // you will have foo.__foo_field_0 = 'email' and row.__foo_value_0 = 'a@b.c',
+                        // so add join/where clause "mdl_foo.email = 'a@b.c'"
+                        clause = `mdl_${m[1]}.${row[x]} = ?`;
+                        have_sub = true;
+                    }else if (parts.length === 2){
+                        // Given defining_attributes('foo') => ["course:c.id"],
+                        // you're linking against a reference in the new table rather than a static value,
+                        // so add join/where clause "mdl_foo.course = c.id"
+                        clause = `${parts[1]} = mdl_${m[1]}.${row[x]} `;
+                    }else{
+                        throw new Error(m[1] + '.' + row[x] + ' is and is not a defining attribute.');
+                    }
+                    if (m[1] === row.objecttable){
+                        where_clause.push(clause);
+                        if (have_sub){
+                            where_subs.push(row[val]);
                         }
-                        if (m[1] === row.objecttable){
-                            where_clause.push(clause);
-                            if (have_sub){
-                                where_subs.push(row[val]);
-                            }
-                        }else{
-                            join_clause.push(clause);
-                            if (have_sub){
-                                join_subs.push(row[val]);
-                            }
+                    }else{
+                        join_clause.push(clause);
+                        if (have_sub){
+                            join_subs.push(row[val]);
                         }
-                        return;
                     }
-                    m = x.match(/__(.*)_operator/);
-                    if (m){
-                        if (m[1] === row.objecttable){
-                            where_op = row[x];
-                        }else{
-                            join_op = row[x];
-                        }
-                        return;
+                    return;
+                }
+                m = x.match(/__(.*)_operator/);
+                if (m){
+                    if (m[1] === row.objecttable){
+                        where_op = row[x];
+                    }else{
+                        join_op = row[x];
                     }
-                });
-
+                    return;
+                }
+            });
+            let where_c = where_clause.join(' '+where_op+' ');
+            let pri_link = make_join(row.objecttable, 'u');
+            let rel_link = make_join(row.objecttable, 'r');
+            let real_link = make_join(row.objecttable, 'a');
+            let course_link = make_join(row.objecttable, 'c');
+            where_c = where_c.replace(pri_link,'').replace(pri_link.replace(' AND ',''),'')
```

Comment:
Is there a better way to do this? We're using the
`x.replace(y, '').replace(y.replace(' AND ', ''), '')`
pattern a lot - what are we really trying to do? Should `y` just not have been added to `where_clause` in the first place?